ocaml/xapi/cert_distrib.ml

```diff
@@ -364,15 +364,32 @@ let collect_pool_certs ~__context ~rpc ~session_id ~map ~from_hosts =
          map cert
      )
 
-let take_and_append n x xs =
-  (* take_and_append 3 10 [1;2;3;4] = [1;2;3;10] *)
-  let rec loop i acc = function
-    | x :: xs when i < n ->
-        loop (i + 1) (x :: acc) xs
-    | _ ->
-        x :: acc |> List.rev
+(* insert an element randomly *between* existing elements of a list.
+   We are enumerating the gaps between elements in a list:
+
+   List:     a b c d
+   Gaps:    0 1 2 3 4
+
+   A list of length n has gaps 0, 1, ...,n. We are selecting a gap
+   between 1 and n-1 to insert an element - so never at the beginning or
+   end if n >= 2. Otherwise we insert at the front.
+
+   This function is not tail recursive, so only apply to short lists *)
+
+let randomly_insert x ys =
+  let gap = Random.int (List.length ys) + 1 in
+  let rec loop n xs =
+    match (n, xs) with
+    | _, [] ->
+        [x]
```

Comment:
If you want to never insert at the end, shouldn't this be the empty list `[]`?

Reply:
This is a matter of definition. If we want to insert something, we have no choice here. I don't think the edge case matters for how we use it but I rather not drop any element.